app/src/main/java/emu/skyline/provider/SkylineDocumentsProvider.kt

```diff
@@ -0,0 +1,307 @@
+import java.io.*
+
+
+class SkylineDocumentsProvider : DocumentsProvider() {
+
+    private val baseDir = File(SkylineApplication.applicationContext().getPublicFilesDir().canonicalPath)
+
+    private val DEFAULT_ROOT_PROJECTION : Array<String> = arrayOf(
+        DocumentsContract.Root.COLUMN_ROOT_ID,
+        DocumentsContract.Root.COLUMN_MIME_TYPES,
+        DocumentsContract.Root.COLUMN_FLAGS,
+        DocumentsContract.Root.COLUMN_ICON,
+        DocumentsContract.Root.COLUMN_TITLE,
+        DocumentsContract.Root.COLUMN_SUMMARY,
+        DocumentsContract.Root.COLUMN_DOCUMENT_ID,
+        DocumentsContract.Root.COLUMN_AVAILABLE_BYTES
+    )
+    private val DEFAULT_DOCUMENT_PROJECTION : Array<String> = arrayOf(
+        DocumentsContract.Document.COLUMN_DOCUMENT_ID,
+        DocumentsContract.Document.COLUMN_MIME_TYPE,
+        DocumentsContract.Document.COLUMN_DISPLAY_NAME,
+        DocumentsContract.Document.COLUMN_LAST_MODIFIED,
+        DocumentsContract.Document.COLUMN_FLAGS,
+        DocumentsContract.Document.COLUMN_SIZE
+    )
+
+    override fun onCreate() : Boolean {
+        return true
+    }
+
+    override fun queryRoots(projection : Array<out String>?) : Cursor {
+        val result = MatrixCursor(projection ?: DEFAULT_ROOT_PROJECTION)
+        val applicationTitle = SkylineApplication.applicationContext().applicationInfo.loadLabel(SkylineApplication.applicationContext().packageManager).toString()
+
+        result.newRow().apply {
+            add(DocumentsContract.Root.COLUMN_ROOT_ID, getDocIdForFile(baseDir))
+            add(DocumentsContract.Root.COLUMN_SUMMARY, null)
+            add(
+                DocumentsContract.Root.COLUMN_FLAGS,
+                DocumentsContract.Root.FLAG_SUPPORTS_CREATE or
+                        DocumentsContract.Root.FLAG_SUPPORTS_SEARCH
+            )
+            add(DocumentsContract.Root.COLUMN_TITLE, applicationTitle)
+            add(DocumentsContract.Root.COLUMN_DOCUMENT_ID, getDocIdForFile(baseDir))
+            add(DocumentsContract.Root.COLUMN_MIME_TYPES, "*/*")
+            add(DocumentsContract.Root.COLUMN_AVAILABLE_BYTES, baseDir.freeSpace)
+            add(DocumentsContract.Root.COLUMN_ICON, null)
+        }
+        return result
+    }
+
+    private fun getDocIdForFile(file : File) : Any? {
+        return file.absolutePath
+    }
+
+    override fun queryDocument(documentId : String?, projection : Array<out String>?) : Cursor {
+        var result = MatrixCursor(projection ?: DEFAULT_DOCUMENT_PROJECTION)
+        result = includeFile(result, documentId, null)
+        return result
+    }
+
+    override fun isChildDocument(parentDocumentId : String?, documentId : String?) : Boolean {
+        return documentId?.startsWith(parentDocumentId!!) ?: false
+    }
+
+    override fun createDocument(parentDocumentId : String?, mimeType : String?, displayName : String) : String? {
+        var newFile = File(parentDocumentId, displayName)
+        var noConflictId = 2
```

Comment:
It is fairly confusing what `noConflictId` is, if this is required at all I'd recommend adding a simple inline comment explaining what it is.